### ISSUE TITLE
Function call counts incorrect, including negative values

### ISSUE THREAD

**Issue by spritchard** [darold/pgcluu#51]
When I run pgcluu_collectd on our system, the pg_stat_user_functions.csv data file contains data about several different functions. When this is run through pgcluu, the call counts on the database-functions page are completely wrong, including negative counts for some functions.

The code in pg_stat_user_functions() seems to assume that there is data on only one function - it subtracts call counts from one row from the next, irrespective of which function the data relates to.

I've modified pgcluu as follows , which seems to work for me:

```
diff --git a/pgcluu b/pgcluu
index b4b2c82..2a34cd2 100755
--- a/pgcluu
+++ b/pgcluu
@@ -3432,6 +3432,7 @@ sub pg_stat_user_functions
 {
        my ($in_dir, $file) = @_;

+       my %save_start_vals = ();
        my @start_vals = ();
        my $tmp_val = 0;
        # Load data from file
@@ -3441,10 +3442,11 @@ sub pg_stat_user_functions
                next if (!&normalize_line(\@data));

                # timestamp | datname | funcid | schemaname | funcname | calls | total_time | self_ti
-               push(@start_vals, @data) if ($#start_vals < 0);
-
                $data[4] = "$data[3].$data[4]";

+               $save_start_vals{$data[4]} = [@data] if not exists $save_start_vals{$data[4]};
+               @start_vals = @{$save_start_vals{$data[4]}};
+
                (($data[5] - $start_vals[5]) < 0) ? $tmp_val = 0 : $tmp_val = ($data[5] - $start_vals
                $all_stat_user_functions{$data[1]}{$data[4]}{calls} += $tmp_val;
                (($data[6] - $start_vals[6]) < 0) ? $tmp_val = 0 : $tmp_val = ($data[6] - $start_vals
@@ -3452,8 +3454,7 @@ sub pg_stat_user_functions
                (($data[7] - $start_vals[7]) < 0) ? $tmp_val = 0 : $tmp_val = ($data[7] - $start_vals
                $all_stat_user_functions{$data[1]}{$data[4]}{self_time} += $tmp_val;

-               @start_vals = ();
-               push(@start_vals, @data);
+               $save_start_vals{$data[4]} = [@data];
        }
        $curfh->close();
 }
```

pgcluu is a great tool!

Steve Pritchard

**Comment by spritchard** [comment]
Thinking about this further, I've realised that the same function could exist on multiple databases. Therefore %save_start_vals should be indexed on database as well as schema.function.

**Comment by darold** [comment]
Thanks for the report Steve, it might now be fixed with commit  b8e4e4a.

Regards,

**Comment by spritchard** [comment]
Looks good. Many thanks.
